source/include/microapp/cs_Microapp.h

```diff
@@ -1,99 +1,115 @@
- * part of the flash memory.
+ * Class that enables the feature to run microapps on the firmware.
+ *
+ * This class:
+ * - Handles control commands (from BLE and UART).
+ * - Keeps up the state of microapps.
+ * - Instantiates and uses classes MicroappStorage and MicroappProtocol.
  */
 class Microapp: public EventListener {
-	private:
-		/**
-		 * Singleton, constructor, also copy constructor, is private.
-		 */
-		Microapp();
-		Microapp(Microapp const&);
-		void operator=(Microapp const &);
-
-		/**
-		 * The last message we have sent. It has a repeat value that will start at "number_of_notifications". If it
-		 * is at zero, it will not be repeated anymore.
-		 */
-		TYPIFY(EVT_MICROAPP) _prevMessage;
-
-		/**
-		 * For the last chunk we have to do some other things. Validate the entire app, etc.
-		 */
-		bool _lastChunk;
-
-		/**
-		 * Local flag to check if microapp is enabled.
-		 */
-		bool _enabled;
-
-		/**
-		 * Local flag to indicate that ram section has been loaded.
-		 */
-		bool _loaded;
-
-		/**
-		 * Debug mode
-		 */
-		bool _debug;
-
-	protected:
-		/**
-		 * Set IPC ram data.
-		 */
-		void setIpcRam();
-
-		/**
-		 * Enable app.
-		 */
-		uint16_t enableApp(bool flag);
-
-		/**
-		 * Check if app is enabled.
-		 */
-		bool isEnabled();
-
-		/**
-		 * Call the loop function
-		 */
-		void callLoop(int & cntr, int & skip);
-
-		/**
-		 * Load ram information, set by microapp.
-		 */
-		uint16_t interpretRamdata();
-	public:
-		static Microapp& getInstance() {
-			static Microapp instance;
-			return instance;
-		}
-
-		/**
-		 * Initialize fstorage. Allocate buffer.
-		 */
-		uint16_t init();
-
-		/**
-		 * The tick function is used to send more notifications than one.
-		 */
-		void tick();
-
-		/**
-		 * Handle incoming events.
-		 */
-		void handleEvent(event_t & event);
-
-		/**
-		 * Handle incoming packet. Returns err_code.
-		 */
-		uint32_t handlePacket(microapp_packet_header_t *packet_stub);
+public:
+	static Microapp& getInstance() {
+		static Microapp instance;
+		return instance;
+	}
+
+	/**
+	 * Initialize storage, and load microapps.
+	 */
+	void init();
+
+private:
+	/**
+	 * Singleton, constructor, also copy constructor, is private.
+	 */
+	Microapp();
+	Microapp(Microapp const&);
+	void operator=(Microapp const &);
+
+	/**
+	 * The state of each microapp.
+	 */
+	microapp_state_t _states[MAX_MICROAPPS];
+
+	/**
+	 * Local flag to indicate that ram section has been loaded.
+	 */
+	bool _loaded = false;
+
+	void loadApps();
+
+	void loadState(uint8_t index);
+
+	/**
+	 * Validates app: compare checksum.
+	 * App state is updated in this call, make sure to store the state afterwards.
+	 */
+	cs_ret_code_t validateApp(uint8_t index);
+
+	/**
+	 * Enables app: checks sdk version.
+	 * App state is updated in this call, make sure to store the state afterwards.
+	 */
+	cs_ret_code_t enableApp(uint8_t index);
+
+	/**
+	 * Start app, if it passed all tests.
```

Comment:
Documentation confuses me: does this perform the tests or not?

Reply:
Changed to "Start app, if it passed all tests according to cached state."

Reply:
"Enables app: checks sdk version." 

Does this mean that at validation there is no check on SDK version?

I think some comments on the intended procedure would be nice:

- There is a validation response for the sending party (over Bluetooth LE) to communicate that the app is written correctly and/or that there is a valid app present at a particular spot.
- There are tests performed by the microapp. The microapp is called, particular functions like setup, loop, and interrupt service routines, are called, etc.
- There is management. Enable/disable an app at particular moments or because the user wants to.

A check for SDK version doesn't seem to belong in a function that enables the app. Or, if things are checked before an app is enabled, I would put these all together in a separate function.

Reply:
Currently, validate only checks data integrity: whether or not the upload succeeded.
The reason that the SDK check is done on enable, is because you can update firmware, and thus change SDK version, while the binary remains the same.
I'm open for discussion though, because i can also think of reasons the SDK check should be done on validation.

Reply:
Maybe an `else nothing happens and ERR_UNSAFE is returned` would clarify things. For most of these methods only the positive result is documented while knowing that nothing happens if it shouldn't is good to know.